presto-main/src/main/java/com/facebook/presto/operator/PagesSpatialIndexSupplier.java

```diff
@@ -0,0 +1,127 @@
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.Session;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.gen.JoinFilterFunctionCompiler;
+import io.airlift.slice.Slice;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import org.locationtech.jts.geom.Envelope;
+import org.locationtech.jts.index.strtree.STRtree;
+
+import javax.annotation.Nullable;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiPredicate;
+
+import static com.facebook.presto.geospatial.GeometryUtils.deserialize;
+import static com.facebook.presto.operator.SyntheticAddress.decodePosition;
+import static com.facebook.presto.operator.SyntheticAddress.decodeSliceIndex;
+import static java.util.Collections.emptyList;
+
+public class PagesSpatialIndexSupplier
+{
+    private static final Envelope ZERO_ENVELOPE = new Envelope(0.0, 0.0, 0.0, 0.0);
+    private static final PagesSpatialIndex EMPTY_INDEX = new PagesSpatialIndex()
+    {
+        @Override
+        public long getEstimatedMemorySizeInBytes()
+        {
+            return 0;
+        }
+
+        @Override
+        public List<Long> findJoinAddresses(int probePosition, Page probe, int probeGeometryChannel, BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest)
+        {
+            return emptyList();
+        }
+
+        @Override
+        public void appendTo(long joinAddress, PageBuilder pageBuilder, int outputChannelOffset)
+        {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    private PagesSpatialIndex index;
+
+    public PagesSpatialIndexSupplier(
+            Session session,
+            LongArrayList addresses,
+            List<Type> types,
+            List<Integer> outputChannels,
+            List<List<Block>> channels,
+            int geometryChannel,
+            Optional<JoinFilterFunctionCompiler.JoinFilterFunctionFactory> filterFunctionFactory)
+    {
+        STRtree rtree = buildRTree(addresses, channels, geometryChannel);
+        if (rtree == null) {
+            index = EMPTY_INDEX;
+        }
+        else {
+            index = new PagesRTreeIndex(session, addresses, types, outputChannels, channels, rtree, filterFunctionFactory);
+        }
+    }
+
+    @Nullable
+    private static STRtree buildRTree(LongArrayList addresses, List<List<Block>> channels, int geometryChannel)
+    {
+        STRtree rtree = new STRtree();
+
+        for (int position = 0; position < addresses.size(); position++) {
+            long pageAddress = addresses.getLong(position);
+            int blockIndex = decodeSliceIndex(pageAddress);
+            int blockPosition = decodePosition(pageAddress);
+
+            Block block = channels.get(geometryChannel).get(blockIndex);
+            if (block.isNull(blockPosition)) {
```

Comment:
I'll add a TODO, but I'm curious why do you think a separate filter will be better. Do you expect it to be faster?

Reply:
I think there is almost always some kind of a filter. There is dedicated filter support in table scan processing, which benefits from lazy-materializing rows.

Also, spatial is a form of cross join, the build side is broadcasted. If rows with null are rare, filtering earlier doesn't matter. If, however, 15% rows have nulls, you save up to 15% of the network (provided there are other non-null columns you would output if join succeeded).

Interestingly though, we could do the same for hash join, but we don't do yet -- https://github.com/prestodb/presto/issues/8796 -- that's why i think "TODO" is fine (or an issue instead of TODO).